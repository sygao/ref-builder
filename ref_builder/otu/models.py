from collections import Counter
from uuid import UUID

from pydantic import UUID4, Field, model_validator

from ref_builder.resources import RepoIsolate, RepoSequence
from ref_builder.resources.models import IsolateModel, OTUModel
from ref_builder.utils import Accession


class IsolateBase(IsolateModel):
    """A class representing an isolate with basic validation."""

    sequences: list[RepoSequence]
    """The isolates sequences."""

    def get_sequence_by_accession(
        self,
        accession: Accession,
    ) -> RepoSequence | None:
        """Get a sequence by its accession.

        Return ``None`` if the sequence is not found.

        :param accession: the accession of the sequence to retrieve
        :return: the sequence with the given accession, or None
        """
        for sequence in self.sequences:
            if sequence.accession == accession:
                return sequence

        return None

    def get_sequence_by_id(self, sequence_id: UUID) -> RepoSequence | None:
        """Get a sequence by its ID.

        Return ``None`` if the sequence is not found.

        :param sequence_id: the ID of the sequence to retrieve
        :return: the sequence with the given ID, or None
        """
        for sequence in self.sequences:
            if sequence.id == sequence_id:
                return sequence

        return None


class Isolate(IsolateBase):
    """A class representing an isolate with full validation."""

    sequences: list[RepoSequence] = Field(min_length=1)
    """The isolates sequences.

    A valid isolate must have at least one sequence.
    """


class OTUBase(OTUModel):
    """A class representing an OTU with basic validation."""

    excluded_accessions: set[str]
    """A set of accessions that should not be retrieved in future fetch operations."""

    isolates: list[IsolateBase]
    """Isolates contained in this OTU."""

    representative_isolate: UUID4 | None
    """The UUID of the representative isolate of this OTU"""

    sequences: list[RepoSequence]
    """Sequences contained in this OTU."""


class OTU(OTUBase):
    """A class representing an OTU with full validation."""

    isolates: list[RepoIsolate] = Field(min_length=1)
    """Isolates contained in this OTU.

    A valid OTU must have at least one isolate.
    """

    representative_isolate: UUID4
    """The UUID of the representative isolate of this OTU.

    A valid OTU must have a representative isolate.
    """

    @model_validator(mode="after")
    def check_excluded_accessions(self) -> "OTU":
        """Ensure that excluded accessions are not in the OTU."""
        if accessions := self.excluded_accessions & {
            sequence.accession.key for sequence in self.sequences
        }:
            raise ValueError(
                f"Excluded accessions found in the OTU: {', '.join(accessions)}"
            )

        return self

    @model_validator(mode="after")
    def check_representative_isolate(self) -> "OTU":
        """Ensure that the default isolate is in the OTU."""
        if self.representative_isolate not in {isolate.id for isolate in self.isolates}:
            raise ValueError("Representative isolate must be in the OTU")

        return self

    @model_validator(mode="after")
    def check_unique_isolate_names(self) -> "OTU":
        """Ensure there are no duplicate isolate names in the OTU."""
        counts = Counter(isolate.name for isolate in self.isolates)

        duplicates = ", ".join(
            [str(name) for name, count in counts.items() if name and count > 1]
        )

        if duplicates:
            raise ValueError(
                f"Isolate names must be unique. Non-unique names: {duplicates}",
            )

        return self

    @model_validator(mode="after")
    def check_isolates_against_plan(self) -> "OTU":
        """Check that all isolates satisfy the OTU's plan.

        TODO: Implement this method.
        """
        return self
