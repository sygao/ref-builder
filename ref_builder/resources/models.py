from pydantic import (
    UUID4,
    BaseModel,
    Field,
    field_serializer,
    field_validator,
)

from ref_builder.models import Molecule
from ref_builder.plan import Plan
from ref_builder.utils import Accession, IsolateName


class SequenceModel(BaseModel):
    """The basic model of a sequence in the OTU."""

    id: UUID4
    """The sequence id."""

    accession: Accession
    """The sequence accession."""

    definition: str
    """The sequence definition."""

    legacy_id: str | None
    """A string based ID carried over from a legacy Virtool reference repository.

    It the sequence was not migrated from a legacy repository, this will be `None`.
    """

    sequence: str = Field(pattern=r"^[ATCGRYKMSWBDHVN]+$")
    """The sequence."""

    segment: str
    """The sequence segment."""

    @field_serializer("accession")
    @classmethod
    def serialize_accession(self, accession: Accession) -> str:
        """Serialize the accession to a string."""
        return str(accession)

    @field_validator("accession", mode="before")
    @classmethod
    def convert_accession(cls, value: Accession | str) -> Accession:
        """Convert the accession to an Accession object."""
        if isinstance(value, Accession):
            return value

        if isinstance(value, str):
            return Accession.from_string(value)

        raise ValueError(f"Invalid type for accession: {type(value)}")


class IsolateModel(BaseModel):
    """Models the basic structure of an isolate in an OTU, sans sequences."""

    id: UUID4
    """The isolate id."""

    legacy_id: str | None
    """A string based ID carried over from a legacy Virtool reference repository.

    It the isolate was not migrated from a legacy repository, this will be `None`.
    """

    name: IsolateName | None
    """The isolate's source name metadata."""

    @field_serializer("name")
    def serialize_name(self, name: IsolateName | None) -> dict[str, str] | None:
        """Serialize the isolate name."""
        if name is None:
            return None

        return {
            "type": name.type,
            "value": name.value,
        }

    @field_validator("name", mode="before")
    @classmethod
    def convert_name(cls, value: dict | IsolateName | None) -> IsolateName | None:
        """Convert the name to an IsolateName object."""
        if value is None:
            return value

        if isinstance(value, IsolateName):
            return value

        if isinstance(value, dict):
            return IsolateName(**value)

        raise ValueError(f"Invalid type for name: {type(value)}")


class OTUModel(BaseModel):
    """Models the basic structure of an OTU, sans isolates."""

    id: UUID4
    """The OTU id."""

    acronym: str
    """The OTU acronym (eg. TMV for Tobacco mosaic virus)."""

    legacy_id: str | None
    """A string based ID carried over from a legacy Virtool reference repository."""

    name: str
    """The name of the OTU (eg. TMV for Tobacco mosaic virus)"""

    molecule: Molecule
    """The type of molecular information contained in this OTU."""

    plan: Plan
    """The schema of the OTU"""

    taxid: int
    """The NCBI Taxonomy id for this OTU."""
