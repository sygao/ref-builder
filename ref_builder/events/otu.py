from pydantic import ConfigDict, UUID4, field_serializer

from ref_builder.events.base import ApplicableEvent, EventData, Event, OTUQuery
from ref_builder.resources.otu import RepoOTU
from ref_builder.models import Molecule
from ref_builder.plan import Plan


class CreateOTUData(EventData):
    """The data for a :class:`CreateOTU` event."""

    id: UUID4
    acronym: str
    legacy_id: str | None
    molecule: Molecule
    name: str
    taxid: int
    plan: Plan


class CreateOTU(Event):
    """An event that creates a new OTU."""

    data: CreateOTUData
    query: OTUQuery

    def apply(self) -> RepoOTU:
        return RepoOTU(
            id=self.query.otu_id,
            acronym=self.data.acronym,
            excluded_accessions=set(),
            isolates=[],
            legacy_id=self.data.legacy_id,
            molecule=self.data.molecule,
            name=self.data.name,
            representative_isolate=None,
            plan=self.data.plan,
            taxid=self.data.taxid,
        )


class CreatePlanData(EventData):
    """The data for a :class:`CreatePlan` event."""

    plan: Plan


class CreatePlan(ApplicableEvent):
    """An event that sets the isolate plan for an OTU."""

    data: CreatePlanData
    query: OTUQuery

    def apply(self, otu: RepoOTU) -> RepoOTU:
        """Apply changed plan to OTU and return."""
        otu.plan = self.data.plan

        return otu


class SetRepresentativeIsolateData(EventData):
    """The data for a :class:`SetReprIsolate` event."""

    isolate_id: UUID4


class SetRepresentativeIsolate(ApplicableEvent):
    """An event that sets the representative isolate for an OTU."""

    data: SetRepresentativeIsolateData
    query: OTUQuery

    def apply(self, otu: RepoOTU) -> RepoOTU:
        """Update the OTU's representative isolate and return."""
        otu.representative_isolate = self.data.isolate_id

        return otu


class UpdateExcludedAccessionsData(EventData):
    """The data for an UpdateAllowedAccessions event."""

    model_config = ConfigDict(use_enum_values=True)

    accessions: set[str]
    action: ExcludedAccessionAction

    @field_serializer("accessions")
    def serialize_accessions(self, accessions: set[str]) -> list[str]:
        return sorted(accessions)


class UpdateExcludedAccessions(ApplicableEvent):
    """An event that changes the OTU excluded accessions collection.

    This event is emitted when Genbank accessions are either
    allowed or disallowed from inclusion in the reference.
    """

    data: UpdateExcludedAccessionsData
    query: OTUQuery

    def apply(self, otu: RepoOTU) -> RepoOTU:
        """Add accession allowance changes to OTU and return."""

        if self.data.action == ExcludedAccessionAction.ALLOW:
            for accession in self.data.accessions:
                otu.excluded_accessions.discard(accession)

        else:
            for accession in self.data.accessions:
                otu.excluded_accessions.add(accession)

        return otu
