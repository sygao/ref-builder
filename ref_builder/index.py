"""An index for improving repository performance."""

import binascii
import sqlite3
from collections.abc import Iterator
from dataclasses import dataclass
from pathlib import Path
from typing import Any
from uuid import UUID

import orjson

from ref_builder.models import OTUMinimal
from ref_builder.resources.otu import RepoOTU


@dataclass
class EventIndexItem:
    """The event IDs associated with an OTU."""

    at_event: int
    """The latest event ID when this cache index was last updated"""

    event_ids: list[int]
    """A list of event IDs"""

    otu_id: UUID
    """The OTU ID this cache index is associated with."""


@dataclass
class Snapshot:
    """A snapshot of an OTU at a specific event."""

    at_event: int
    """The event at which the snapshot was taken."""

    otu: RepoOTU
    """The OTU that was snapshotted."""


class Index:
    """An index for rapid access to repository data.

    1. Get OTU IDs by legacy ID, name, or taxonomy ID.
    2. Save and load complete snapshots of OTUs.
    3. Get the event IDS associated with an OTU.

    """

    def __init__(self, path: Path) -> None:
        """Initialize the index."""
        self.path = path
        """The path to the index file."""

        self.path.parent.mkdir(exist_ok=True, parents=True)

        self.con = sqlite3.connect(path, isolation_level=None)
        self.con.execute("PRAGMA journal_mode = WAL")
        self.con.execute("PRAGMA synchronous = NORMAL")

        self.con.execute(
            """
            CREATE TABLE IF NOT EXISTS at_event (
                id INTEGER PRIMARY KEY,
                at_event INTEGER
            );
            """,
        )

        self.con.execute(
            """
            CREATE TABLE IF NOT EXISTS events (
                at_event INTEGER,
                event_id INTEGER PRIMARY KEY,
                otu_id TEXT
            );
            """,
        )

        self.con.execute(
            """
            CREATE TABLE IF NOT EXISTS otus (
                id TEXT PRIMARY KEY,
                acronym TEXT,
                at_event INTEGER,
                legacy_id TEXT,
                name TEXT,
                otu JSONB,
                taxid INTEGER
            );
            """,
        )

        self.con.execute(
            """
            CREATE TABLE IF NOT EXISTS sequences (
                id TEXT PRIMARY KEY,
                crc TEXT,
                otu_id TEXT,
                sequence TEXT
            );
            """,
        )

        for table, column in [
            ("events", "otu_id"),
            ("otus", "id"),
            ("otus", "legacy_id"),
            ("otus", "name"),
            ("otus", "taxid"),
            ("sequences", "otu_id"),
            ("sequences", "crc"),
        ]:
            self.con.execute(
                f"""
                CREATE INDEX IF NOT EXISTS {table}_{column} ON {table} ({column});
                """,
            )

        self.con.commit()

    @property
    def otu_ids(self) -> set[UUID]:
        """A list of all OTUs tracked in the index."""
        return {UUID(row[0]) for row in self.con.execute("SELECT id FROM otus")}

    def add_event_id(self, event_id: int, otu_id: UUID) -> None:
        """Add a new event ID and associated it with a given OTU ID.

        This allows fast lookup of event IDs for a given OTU ID.

        :param event_id: the event ID to add
        :param otu_id: the OTU ID to add the event ID to
        """
        self.con.execute(
            """
            INSERT INTO events (event_id, otu_id)
            VALUES (?, ?)
            """,
            (
                event_id,
                str(otu_id),
            ),
        )

    def get_event_ids_by_otu_id(self, otu_id: UUID) -> EventIndexItem | None:
        """Get a list of event IDs for an OTU."""
        res = self.con.execute(
            """
            SELECT event_id
            FROM events
            WHERE otu_id = ?
            ORDER BY event_id
            """,
            (str(otu_id),),
        )

        event_ids = [row[0] for row in res]

        if event_ids:
            return EventIndexItem(
                event_ids[-1],
                event_ids,
                otu_id,
            )

        return None

    def get_id_by_legacy_id(self, legacy_id: str) -> UUID | None:
        """Get an OTU ID by its legacy ID."""
        cursor = self.con.execute(
            "SELECT id FROM otus WHERE legacy_id = ?",
            (legacy_id,),
        )

        if result := cursor.fetchone():
            return UUID(result[0])

        return None

    def get_id_by_name(self, name: str) -> UUID | None:
        """Get an OTU ID by its name."""
        cursor = self.con.execute(
            "SELECT id FROM otus WHERE name = ?",
            (name,),
        )

        if result := cursor.fetchone():
            return UUID(result[0])

        return None

    def get_id_by_taxid(self, taxid: int) -> UUID | None:
        """Get an OTU ID by its taxonomy ID."""
        cursor = self.con.execute(
            "SELECT id FROM otus WHERE taxid = ?",
            (taxid,),
        )

        if result := cursor.fetchone():
            return UUID(result[0])

        return None

    def delete_otu(self, otu_id: UUID) -> None:
        """Remove an OTU from the index.

        This happens rarely, so we just rewrite the whole index.

        :param otu_id: the ID of the OTU to remove.
        """
        self.con.execute(
            "DELETE FROM events WHERE otu_id = ?",
            (str(otu_id),),
        )

        self.con.execute(
            "DELETE FROM otus WHERE id = ?",
            (str(otu_id),),
        )

        self.con.execute(
            "DELETE FROM sequences WHERE otu_id = ?",
            (str(otu_id),),
        )

        self.con.commit()

    def iter_minimal_otus(self) -> Iterator[OTUMinimal]:
        """Iterate over minimal representations of all OTUs in the index."""
        cursor = self.con.execute(
            "SELECT acronym, id, legacy_id, name, taxid FROM otus ORDER BY name",
        )

        for row in cursor:
            yield OTUMinimal(
                acronym=row[0],
                id=UUID(row[1]),
                legacy_id=row[2],
                name=row[3],
                taxid=row[4],
            )

    def load_snapshot(self, otu_id: UUID) -> Snapshot | None:
        """Load an OTU snapshot."""
        cursor = self.con.execute(
            "SELECT at_event, otu FROM otus WHERE id = ?",
            (str(otu_id),),
        )

        result = cursor.fetchone()

        if not result:
            return None

        at_event, otu_json = result

        otu = orjson.loads(otu_json)

        sequence_ids = [
            str(sequence["id"])
            for isolate in otu["isolates"]
            for sequence in isolate["sequences"]
        ]

        # Fetch all sequences in a single query
        placeholders = ",".join("?" for _ in sequence_ids)

        sequence_map = dict(
            self.con.execute(
                f"SELECT id, sequence FROM sequences WHERE id IN ({placeholders})",
                sequence_ids,
            ).fetchall(),
        )

        # Update the data structure and check for missing sequences
        missing_sequences = []

        for isolate in otu["isolates"]:
            for sequence in isolate["sequences"]:
                sequence_id = str(sequence["id"])

                if sequence_id in sequence_map:
                    sequence["sequence"] = sequence_map[sequence_id]
                else:
                    missing_sequences.append(sequence_id)

        # Raise an error if any sequences are missing
        if missing_sequences:
            raise ValueError(f"Sequences not found: {', '.join(missing_sequences)}")

        return Snapshot(
            at_event=at_event,
            otu=RepoOTU.model_validate(otu),
        )

    def upsert_otu(self, otu: RepoOTU, at_event: int) -> None:
        """Update the index based on a list of OTUs.

        OTUs that aren't already in the index will be added, and those that are will be
        updated.

        Only sequences that have changed will be updated.

        """
        sequence_ids = {
            str(seq.id) for isolate in otu.isolates for seq in isolate.sequences
        }

        otu_dict = otu.model_dump(mode="json")

        placeholders = ",".join("?" for _ in sequence_ids)

        # Delete any sequences that are no longer in the OTU.
        self.con.execute(
            f"""
            DELETE FROM sequences WHERE otu_id = ? AND NOT id IN ({ placeholders });
            """,
            (str(otu.id), *sequence_ids),
        )

        batch = []

        # Insert or update the sequences.
        for isolate in otu_dict["isolates"]:
            for sequence in isolate["sequences"]:
                crc = _calculate_crc32(sequence["sequence"])
                batch.append(
                    (
                        str(sequence["id"]),
                        crc,
                        str(otu.id),
                        sequence["sequence"],
                    ),
                )

        # Perform batch insert
        self.con.executemany(
            """
            INSERT OR REPLACE INTO sequences (id, crc, otu_id, sequence)
            VALUES (?, ?, ?, ?)
            """,
            batch,
        )

        # Update only if CRC has changed
        self.con.executemany(
            """
            UPDATE sequences
            SET crc = ?, otu_id = ?, sequence = ?
            WHERE id = ? AND crc != ?
            """,
            [(crc, otu_id, seq, seq_id, crc) for seq_id, crc, otu_id, seq in batch],
        )

        self.con.execute(
            """
            INSERT OR REPLACE INTO otus (id, acronym, at_event, legacy_id, name, otu, taxid)
            VALUES (?, ?,?, ?, ?, ?, ?)
            """,
            (
                str(otu.id),
                otu.acronym,
                at_event,
                otu.legacy_id,
                otu.name,
                orjson.dumps(otu_dict),
                otu.taxid,
            ),
        )

        self.con.commit()


def _calculate_crc32(sequence: str) -> str:
    """Calculate the CRC32 checksum of a sequence.

    :param sequence: the sequence to calculate the checksum of.
    :return: the CRC32 checksum as a hexadecimal string.

    """
    crc = binascii.crc32(sequence.encode("utf-8"))

    # Convert CRC as a hexadecimal string.
    return hex(crc & 0xFFFFFFFF)[2:].zfill(8)


def _default_json(obj: Any) -> list:
    """Cast sets to lists.

    This function is used as the default argument to `orjson.dumps` to handle sets.

    :param obj: the object to serialize
    :return: the serialized object
    """
    if isinstance(obj, set):
        return list(obj)

    raise TypeError(f"Object of type {obj.__class__.__name__} is not JSON serializable")
