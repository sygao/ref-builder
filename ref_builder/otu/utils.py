import re
from collections import defaultdict
from enum import StrEnum
from uuid import UUID

import structlog

from ref_builder.models import Molecule
from ref_builder.ncbi.models import NCBIGenbank
from ref_builder.plan import (
    MonopartitePlan,
    MultipartitePlan,
    Segment,
    SegmentRule,
)
from ref_builder.utils import Accession, IsolateName, IsolateNameType

logger = structlog.get_logger("otu.utils")


class DeleteRationale(StrEnum):
    """Default strings delineating reasons for resource deletion."""

    USER = "Requested by user"
    REFSEQ = "Superceded by RefSeq"


class RefSeqConflictError(ValueError):
    """Raised when a potential RefSeq replacement is found."""

    def __init__(
        self,
        message: str,
        isolate_id: UUID,
        isolate_name: IsolateName,
        accessions: list[str],
    ):
        super().__init__(message)

        self.isolate_id = isolate_id

        self.isolate_name = isolate_name

        self.accessions = accessions


def create_segments_from_records(
    records: list[NCBIGenbank], rule: SegmentRule, length_tolerance: float
) -> list[Segment]:
    """Return a list of SegmentPlans."""
    segments = []

    for record in sorted(records, key=lambda record: record.accession):
        if not record.source.segment:
            raise ValueError("No segment name found for multipartite OTU segment.")

        segments.append(
            Segment.from_record(
                record=record,
                length_tolerance=length_tolerance,
                required=rule,
            ),
        )

    return segments


def create_isolate_plan_from_records(
    records: list[NCBIGenbank],
    length_tolerance: float,
    segments: list[Segment] | None = None,
) -> MonopartitePlan | MultipartitePlan | None:
    """Return a plan from a list of records representing an isolate."""
    if len(records) == 1:
        return MonopartitePlan.new(
            length=len(records[0].sequence),
            length_tolerance=length_tolerance,
        )

    binned_records = group_genbank_records_by_isolate(records)
    if len(binned_records) > 1:
        logger.fatal(
            "More than one isolate found. Cannot create schema automatically.",
            bins=binned_records,
        )
        return None

    if segments is not None:
        return MultipartitePlan.new(segments=segments)

    segments = create_segments_from_records(
        records,
        rule=SegmentRule.REQUIRED,
        length_tolerance=length_tolerance,
    )

    if segments:
        return MultipartitePlan.new(segments=segments)

    return None


def group_genbank_records_by_isolate(
    records: list[NCBIGenbank],
) -> dict[IsolateName, dict[Accession, NCBIGenbank]]:
    """Indexes Genbank records by isolate name."""
    isolates = defaultdict(dict)

    for record in records:
        try:
            isolate_name = _get_isolate_name(record)

            if isolate_name is None:
                logger.debug(
                    "RefSeq record does not contain sufficient source data "
                    " for automatic inclusion. Add this record manually.",
                    accession=record.accession,
                    definition=record.definition,
                    source_data=record.source,
                )
                continue

            versioned_accession = Accession.from_string(record.accession_version)
            isolates[isolate_name][versioned_accession] = record

        except ValueError:
            continue

    return isolates


def parse_refseq_comment(comment: str) -> tuple[str, str]:
    """Parse a standard RefSeq comment."""
    if not comment:
        raise ValueError("Empty comment")

    refseq_pattern = re.compile(r"^(\w+ REFSEQ): [\w ]+. [\w ]+ ([\w]+).")

    match = refseq_pattern.search(comment)

    if match:
        return match.group(1), match.group(2)

    raise ValueError("Invalid RefSeq comment")


def get_molecule_from_records(records: list[NCBIGenbank]) -> Molecule:
    """Return relevant molecule metadata from one or more records."""
    if not records:
        raise IndexError("No records given")

    rep_record = None
    for record in records:
        if record.refseq:
            rep_record = record
            break
    if rep_record is None:
        rep_record = records[0]

    return Molecule.model_validate(
        {
            "strandedness": rep_record.strandedness.value,
            "type": rep_record.moltype.value,
            "topology": rep_record.topology.value,
        },
    )


def check_sequence_length(sequence: str, segment_length: int, tolerance: float) -> bool:
    if len(sequence) < segment_length * (1.0 - tolerance) or len(
        sequence
    ) > segment_length * (1.0 + tolerance):
        return False

    return True


def _get_isolate_name(record: NCBIGenbank) -> IsolateName | None:
    """Get the isolate name from a Genbank record."""
    if record.source.model_fields_set.intersection(
        {IsolateNameType.ISOLATE, IsolateNameType.STRAIN, IsolateNameType.CLONE},
    ):
        for source_type in IsolateNameType:
            if source_type in record.source.model_fields_set:
                return IsolateName(
                    type=IsolateNameType(source_type),
                    value=record.source.model_dump()[source_type],
                )

    if record.refseq:
        return None

    raise ValueError("Record does not contain sufficient source data for inclusion.")
