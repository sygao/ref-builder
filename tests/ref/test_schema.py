import pytest
from pydantic import TypeAdapter
from syrupy import SnapshotAssertion
from syrupy.filters import props

from ref_builder.ncbi.client import NCBIClient
from ref_builder.ncbi.models import NCBISourceMolType
from ref_builder.otu.utils import create_isolate_plan_from_records
from ref_builder.plan import (
    Plan,
    MonopartitePlan,
    MultipartitePlan,
    SegmentName,
    get_multipartite_segment_name,
    parse_segment_name,
)
from tests.fixtures.factories import NCBIGenbankFactory, NCBISourceFactory


plan_typer = TypeAdapter(Plan)


class TestPlan:
    """Test differentiated isolate plans."""

    def test_create_monopartite_plan_from_records(
        self,
        scratch_ncbi_client: NCBIClient,
        snapshot: SnapshotAssertion,
    ):
        """Test the creation of a monopartite plan from a given record."""
        records = scratch_ncbi_client.fetch_genbank_records(["NC_024301"])
        auto_plan = create_isolate_plan_from_records(records, length_tolerance=0.03)

        assert type(auto_plan) is MonopartitePlan

        assert auto_plan.model_dump() == snapshot(exclude=props("id"))

    def test_create_multipartite_plan_from_records(
        self,
        scratch_ncbi_client: NCBIClient,
        snapshot: SnapshotAssertion,
    ):
        """Test the creation of a multipartite plan from a set of records
        representing segments.
        """
        records = scratch_ncbi_client.fetch_genbank_records(
            [
                "NC_010314",
                "NC_010315",
                "NC_010316",
                "NC_010317",
                "NC_010318",
                "NC_010319",
            ]
        )
        auto_plan = create_isolate_plan_from_records(records, length_tolerance=0.03)

        assert type(auto_plan) is MultipartitePlan

        assert auto_plan.model_dump() == snapshot(exclude=props("id"))

    @pytest.mark.parametrize(
        ("accessions", "plan_type"),
        [
            (["NC_024301"], MonopartitePlan),
            (
                [
                    "NC_010314",
                    "NC_010315",
                    "NC_010316",
                    "NC_010317",
                    "NC_010318",
                    "NC_010319",
                ],
                MultipartitePlan,
            ),
        ],
    )
    def test_serialize_plan(
        self,
        accessions: list[str],
        plan_type: MonopartitePlan | MultipartitePlan,
        scratch_ncbi_client: NCBIClient,
        snapshot: SnapshotAssertion,
    ):
        records = scratch_ncbi_client.fetch_genbank_records(accessions)

        auto_plan = create_isolate_plan_from_records(records, length_tolerance=0.03)

        assert auto_plan.model_dump() == snapshot(exclude=props("id"))

        auto_plan_json = auto_plan.model_dump_json()

        reconstituted_plan = plan_typer.validate_json(auto_plan_json)

        assert type(reconstituted_plan) is plan_type

        assert reconstituted_plan == auto_plan


class TestSegmentNameParser:
    """Test segment name normalization."""

    @pytest.mark.parametrize(
        ("expected_result", "test_strings"),
        [
            (SegmentName(prefix="DNA", key="A"), ["DNA A", "DNA_A", "DNA-A"]),
            (SegmentName(prefix="RNA", key="BN"), ["RNA BN", "RNA_BN", "RNA-BN"]),
            (SegmentName(prefix="DNA", key="U3"), ["DNA U3", "DNA U3", "DNA-U3"]),
        ],
    )
    def test_ok(self, expected_result: str, test_strings: list[str]):
        """Test that parse_segment_name() correctly parses prefix-key segment names."""
        assert (
            parse_segment_name(test_strings[0])
            == parse_segment_name(test_strings[1])
            == parse_segment_name(test_strings[2])
            == expected_result
        )

    @pytest.mark.parametrize(
        "fail_case",
        ["", "*V/", "51f9a0bc-7b3b-434f-bf4c-f7abaa015b8d"],
    )
    def test_fail(self, fail_case: str):
        """Test that parse_segment_name() returns expected ValueError for bad input."""
        with pytest.raises(
            ValueError,
            match=r"\s* is not a valid segment name$",
        ):
            parse_segment_name(fail_case)

    @pytest.mark.parametrize(
        ("expected_result", "test_name", "test_moltype"),
        [
            (SegmentName(prefix="DNA", key="A"), "A", NCBISourceMolType.GENOMIC_DNA),
            (SegmentName(prefix="RNA", key="BN"), "BN", NCBISourceMolType.GENOMIC_RNA),
            (SegmentName(prefix="DNA", key="U3"), "U3", NCBISourceMolType.GENOMIC_DNA),
        ],
    )
    def test_parse_from_record_ok(
        self,
        expected_result: SegmentName,
        test_name: str,
        test_moltype: NCBISourceMolType,
    ):
        """Test that a full segment name can be parsed
        when the source table segment field contains a key with no prefix.
        """
        dummy_record = NCBIGenbankFactory.build(
            source=NCBISourceFactory.build(
                mol_type=test_moltype,
                segment=test_name,
            ),
        )

        assert get_multipartite_segment_name(dummy_record) == expected_result

    @pytest.mark.parametrize(
        ("test_name", "test_moltype"),
        [
            ("", NCBISourceMolType.GENOMIC_DNA),
            ("V#", NCBISourceMolType.GENOMIC_RNA),
            ("51f9a0bc-7b3b-434f-bf4c-f7abaa015b8d", NCBISourceMolType.GENOMIC_DNA),
        ],
    )
    def test_parse_from_record_fail(
        self,
        test_name: str,
        test_moltype: NCBISourceMolType,
    ):
        """Test that get_multipartite_segment_name()
        does not return an invalid segment name.
        """
        dummy_record = NCBIGenbankFactory.build(
            source=NCBISourceFactory.build(
                mol_type=test_moltype,
                segment=test_name,
            ),
        )

        with pytest.raises(ValueError, match=r"\s* is not a valid segment name$"):
            get_multipartite_segment_name(dummy_record)
