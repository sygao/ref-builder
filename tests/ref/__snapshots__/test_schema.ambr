# serializer version: 1
# name: TestPlan.test_create_monopartite_plan_from_records
  dict({
    'length': 7424,
    'length_tolerance': 0.03,
    'name': None,
    'plan_type': 'monopartite',
  })
# ---
# name: TestPlan.test_create_multipartite_plan_from_records
  dict({
    'plan_type': 'multipartite',
    'segments': list([
      dict({
        'length': 1090,
        'length_tolerance': 0.03,
        'name': dict({
          'key': 'N',
          'prefix': 'DNA',
        }),
        'required': 'required',
      }),
      dict({
        'length': 1057,
        'length_tolerance': 0.03,
        'name': dict({
          'key': 'U3',
          'prefix': 'DNA',
        }),
        'required': 'required',
      }),
      dict({
        'length': 1087,
        'length_tolerance': 0.03,
        'name': dict({
          'key': 'S',
          'prefix': 'DNA',
        }),
        'required': 'required',
      }),
      dict({
        'length': 1074,
        'length_tolerance': 0.03,
        'name': dict({
          'key': 'M',
          'prefix': 'DNA',
        }),
        'required': 'required',
      }),
      dict({
        'length': 1015,
        'length_tolerance': 0.03,
        'name': dict({
          'key': 'C',
          'prefix': 'DNA',
        }),
        'required': 'required',
      }),
      dict({
        'length': 1099,
        'length_tolerance': 0.03,
        'name': dict({
          'key': 'R',
          'prefix': 'DNA',
        }),
        'required': 'required',
      }),
    ]),
  })
# ---
# name: TestPlan.test_serialize_plan[accessions0-MonopartitePlan]
  dict({
    'length': 7424,
    'length_tolerance': 0.03,
    'name': None,
    'plan_type': 'monopartite',
  })
# ---
# name: TestPlan.test_serialize_plan[accessions1-MultipartitePlan]
  dict({
    'plan_type': 'multipartite',
    'segments': list([
      dict({
        'length': 1090,
        'length_tolerance': 0.03,
        'name': dict({
          'key': 'N',
          'prefix': 'DNA',
        }),
        'required': 'required',
      }),
      dict({
        'length': 1057,
        'length_tolerance': 0.03,
        'name': dict({
          'key': 'U3',
          'prefix': 'DNA',
        }),
        'required': 'required',
      }),
      dict({
        'length': 1087,
        'length_tolerance': 0.03,
        'name': dict({
          'key': 'S',
          'prefix': 'DNA',
        }),
        'required': 'required',
      }),
      dict({
        'length': 1074,
        'length_tolerance': 0.03,
        'name': dict({
          'key': 'M',
          'prefix': 'DNA',
        }),
        'required': 'required',
      }),
      dict({
        'length': 1015,
        'length_tolerance': 0.03,
        'name': dict({
          'key': 'C',
          'prefix': 'DNA',
        }),
        'required': 'required',
      }),
      dict({
        'length': 1099,
        'length_tolerance': 0.03,
        'name': dict({
          'key': 'R',
          'prefix': 'DNA',
        }),
        'required': 'required',
      }),
    ]),
  })
# ---
